src/main/resources/com/google/api/codegen/java/message.snip

```diff
@@ -6,18 +6,37 @@
 
   @@Generated("by GAPIC")
   @@BetaApi
-  public final class {@apiFile.schema.typeName} implements Serializable {
-    {@members(apiFile.schema)}
-    {@constructor(apiFile.schema)}
-    {@getters(apiFile.schema)}
-    {@newBuilder(apiFile.schema)}
-    {@builder(apiFile.schema)}
-    {@toString(apiFile.schema)}
-    {@equalsMethod(apiFile.schema)}
-    {@hashCode(apiFile.schema)}
+  {@schemaClassDef(apiFile.schema)}
+@end
+
+@private schemaClassDef(schema)
+  public final class {@schema.innerTypeName} implements Serializable {
+    {@members(schema)}
+
+    {@constructor(schema)}
+
+    {@getters(schema)}
+    {@innerSchemas(schema)}
+    {@newBuilder(schema)}
+
+    {@builder(schema)}
+
+    {@toString(schema)}
+
+    {@equalsMethod(schema)}
+
+    {@hashCode(schema)}
   }
 @end
 
+@private innerSchemas(schema)
+  @join property : schema.properties
+    @if {@property.properties}
+      {@schemaClassDef(property)}
```

Comment:
I'm concerned about the deeply nested classes - it's not typical class design. I don't actually have a better suggestion though because of the potential problems with name clashes if nested classes aren't used.